### PR TITLE
Added better support for XML types into Flow

### PR DESCRIPTION
**Pull request by norberttech**
<h2>Change Log</h2>
<div id="change-log">
  <h4>Added</h4>
  <ul id="added">
    <li>XMLEntry</li>
    <li>XMLNodeEntry</li>
    <li>ref('...')->xpath('...') - for extracting specific nodes from XMLEntry</li>
    <li>ref('...')->domNodeAttribute('...') - for extracting value of attribute</li>
    <li>ref('...')->domNodeValue('...') - for extracting value of node</li>
  </ul> 
  <h4>Fixed</h4>  
  <ul id="fixed">
    
  </ul>
  <h4>Changed</h4>
  <ul id="changed">
     <li>XMLReaderExtractor is now returning XMLEntry type instead of casting XML's to array</li>
  </ul>  
  <h4>Removed</h4>
  <ul id="removed">
    
  </ul>
  <h4>Deprecated</h4>
  <ul id="deprecated">
    
  </ul>  
  <h4>Security</h4> 
  <ul id="security">
    
  </ul>     
</div>
<hr/>

<h2>Description</h2>

The goal of this PR is to make XMLReaderExtractor consistent with Extractors architecture. 
Originally it was reading XML and transforming NodeElements into arrays, this was heavily inspired by Spark however it was adding transformation responsibility to the Extractor. 

This PR changes that (which is a minor BC break) and from now XMLReaderExtractor is throwing XMLEntry which can be later transformer using dedicated XML transformers. 

Following code explains how this can be done: 

<details>
  <summary>XML</summary>

```XML

<Salaries>
    <Month name="January">
        <Department name="HR">
            <TotalSalary>71883</TotalSalary>
        </Department>
        <Department name="Engineering">
            <TotalSalary>192644</TotalSalary>
        </Department>
        <Department name="Finance">
            <TotalSalary>174187</TotalSalary>
        </Department>
        <Department name="Marketing">
            <TotalSalary>179932</TotalSalary>
        </Department>
        <Department name="Sales">
            <TotalSalary>52056</TotalSalary>
        </Department>
    </Month>
    <Month name="February">
        <Department name="HR">
            <TotalSalary>102342</TotalSalary>
        </Department>
        <Department name="Engineering">
            <TotalSalary>111102</TotalSalary>
        </Department>
        <Department name="Finance">
            <TotalSalary>81938</TotalSalary>
        </Department>
        <Department name="Marketing">
            <TotalSalary>132202</TotalSalary>
        </Department>
        <Department name="Sales">
            <TotalSalary>173225</TotalSalary>
        </Department>
    </Month>
    <Month name="March">
        <Department name="HR">
            <TotalSalary>79619</TotalSalary>
        </Department>
        <Department name="Engineering">
            <TotalSalary>99387</TotalSalary>
        </Department>
        <Department name="Finance">
            <TotalSalary>198847</TotalSalary>
        </Department>
        <Department name="Marketing">
            <TotalSalary>50550</TotalSalary>
        </Department>
        <Department name="Sales">
            <TotalSalary>98212</TotalSalary>
        </Department>
    </Month>
    <Month name="April">
        <Department name="HR">
            <TotalSalary>69721</TotalSalary>
        </Department>
        <Department name="Engineering">
            <TotalSalary>151826</TotalSalary>
        </Department>
        <Department name="Finance">
            <TotalSalary>158168</TotalSalary>
        </Department>
        <Department name="Marketing">
            <TotalSalary>111872</TotalSalary>
        </Department>
        <Department name="Sales">
            <TotalSalary>172334</TotalSalary>
        </Department>
    </Month>
    <Month name="May">
        <Department name="HR">
            <TotalSalary>174220</TotalSalary>
        </Department>
        <Department name="Engineering">
            <TotalSalary>164086</TotalSalary>
        </Department>
        <Department name="Finance">
            <TotalSalary>104257</TotalSalary>
        </Department>
        <Department name="Marketing">
            <TotalSalary>105817</TotalSalary>
        </Department>
        <Department name="Sales">
            <TotalSalary>145490</TotalSalary>
        </Department>
    </Month>
    <Month name="June">
        <Department name="HR">
            <TotalSalary>127383</TotalSalary>
        </Department>
        <Department name="Engineering">
            <TotalSalary>52592</TotalSalary>
        </Department>
        <Department name="Finance">
            <TotalSalary>71732</TotalSalary>
        </Department>
        <Department name="Marketing">
            <TotalSalary>165083</TotalSalary>
        </Department>
        <Department name="Sales">
            <TotalSalary>85138</TotalSalary>
        </Department>
    </Month>
</Salaries>
```
</details>

Code: 
```php
<?php

(new Flow())
    ->read(XML::from(__FLOW_DATA__ . '/salaries.xml'))
    ->withEntry('months', ref('row')->xpath('/Salaries/Month'))
    ->withEntry('month', ref('months')->expand())
    ->withEntry('month_name', ref('month')->domNodeAttribute('name'))
    ->withEntry('departments', ref('month')->xpath('/Month/Department'))
    ->withEntry('department', ref('departments')->expand())
    ->withEntry('department_name', ref('department')->domNodeAttribute('name'))
    ->withEntry('department_salary', ref('department')->xpath('/Department/TotalSalary')->domNodeValue())
    ->drop('row', 'months', 'month', 'departments', 'department')
    ->groupBy(ref('month_name'))
    ->aggregate(Aggregation::sum(ref('department_salary')))
    ->rename('department_salary_sum', 'total_monthly_salaries')
    ->write(To::output(false))
    ->run();
```

Output: 
```shell
Reading XML dataset...
+------------+------------------------+
| month_name | total_monthly_salaries |
+------------+------------------------+
|    January |                 670702 |
|   February |                 600809 |
|      March |                 526615 |
|      April |                 663921 |
|        May |                 693870 |
|       June |                 501928 |
+------------+------------------------+
6 rows
```

### CODE REVIEW

File: src/core/etl/src/Flow/ETL/Row/Reference/Expression/Cast/XMLConverter.php [hunk 1]
```diff
@@ -0,0 +1,92 @@
...
+
+        $nodeNames = [];
+        /** @var \DOMElement $childNode */
+        foreach ($element->childNodes as $childNode) {
+            if ($childNode->nodeType === XML_ELEMENT_NODE) {
+                $nodeNames[] = $childNode->nodeName;
+            }
+        }
+
+        if (!\count($nodeNames) || \count($nodeNames) === 1) {
```
Comment (stloyd) [inline src/core/etl/src/Flow/ETL/Row/Reference/Expression/Cast/XMLConverter.php]:
```suggestion
        if (\count($nodeNames) <= 1) {
```

File: src/core/etl/src/Flow/ETL/Rows.php [hunk 2]
```diff
@@ -177,6 +177,31 @@ public function entries() : array
...
+                if ($comparator->equals($row, $otherRow)) {
+                    $found = true;
+
+                    break;
+                }
+            }
+
+            if (!$found) {
+                return false;
+            }
```
Comment (stloyd) [inline src/core/etl/src/Flow/ETL/Rows.php]: Probably can be simplified to:
```suggestion
            foreach ($rows->rows as $otherRow) {
                if ($comparator->equals($row, $otherRow)) {
                    return true;
                }
            }

            return false;
```

File: .php-cs-fixer.php
Comment (stloyd) [inline .php-cs-fixer.php:1]: I guess it was not supposed to make this file executable.

File: src/core/etl/src/Flow/ETL/Rows.php (hunk 2 above)
Comment (norberttech) [reply src/core/etl/src/Flow/ETL/Rows.php]: I think I'm going to revert that change. I did it when I got stuck with comparing XML's, but it turned out that the process of canonicalization of XMLs does not remove whitespaces and line breakers between nodes. 
That method could still be very useful but now when I think about it deserves a separate pull request. However, your suggestion looks valid (but only tests can confirm that), will try to remember about it 😁

File: src/core/etl/src/Flow/ETL/Row/Reference/Expression/Cast.php [hunk 3]
```diff
@@ -46,12 +46,42 @@ public function eval(Row $row) : mixed
...
+        return (array) $data;
+    }
+
+    private function toXML(mixed $value) : \DOMDocument
+    {
+        if (\is_string($value)) {
+            $doc = new \DOMDocument();
+            $loaded = @$doc->load($value);
+
+            if (!$loaded) {
```
Comment (JacekPolanski) [inline src/core/etl/src/Flow/ETL/Row/Reference/Expression/Cast.php]: `$loaded` is useless here

```suggestion
            if (!@$doc->load($value)) {
```

File: src/core/etl/src/Flow/ETL/Row/Reference/Expression/Cast/XMLConverter.php [hunk 4]
```diff
@@ -0,0 +1,92 @@
...
+namespace Flow\ETL\Row\Reference\Expression\Cast;
+
+final class XMLConverter
+{
+    /**
+     * @param \DOMDocument $document
+     *
+     * @return array<mixed>
+     */
+    public function toArray(\DOMDocument $document) : array
```
Comment (JacekPolanski) [inline src/core/etl/src/Flow/ETL/Row/Reference/Expression/Cast/XMLConverter.php:12]: This PHPDoc is not useless here?

File: src/core/etl/src/Flow/ETL/Row/Reference/Expression/Cast/XMLConverter.php [hunk 5]
```diff
@@ -0,0 +1,92 @@
...
+    {
+        $xmlArray = [];
+
+        if ($document->hasChildNodes()) {
+            $children = $document->childNodes;
+
+            foreach ($children as $child) {
```
Comment (JacekPolanski) [inline src/core/etl/src/Flow/ETL/Row/Reference/Expression/Cast/XMLConverter.php]: IMO `$children` variable is useless here and `$child` should be renamed to `$node`

**Comment by norberttech** [comment]
@stloyd @JacekPolanski thanks for review 🙏

**Pull request merged**